pkg/dio/writer.go

```diff
@@ -49,34 +50,41 @@ func (t *multiWriter) Write(p []byte) (int, error) {
 	}
 
 	t.mutex.Lock()
+
+	t.waitGroup.Add(1)
```

Comment:
Any in-flight writes should be protected - any higher level behaviour needs to be enforced by callers.